### PR TITLE
RevenueBuyBack

### PR DESCRIPTION
**Pull request by naddison36**
* New `RevenueBuyBack` contract that uses mUSD and mBTC protocol revenue to buy MTA rewards and sent to stakers.
* Changed `BasicRewardsForwarder` to be `Ownable` so `setEndRecipient` can be called by the TreasuryDAO or a third party. ie the owner.

### CODE REVIEW

File: contracts/buy-and-make/RevenueBuyBack.sol [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+     * @notice donates purchased rewards, eg MTA, to staking contracts via the Emissions Controller.
+     */
+    function donateRewards() external keeperOrGovernor {
+        // STEP 1 - Get the voting power of the staking contracts
+        uint256 numberStakingContracts = stakingDialIds.length;
+        uint256[] memory votingPower = new uint256[](numberStakingContracts);
+        uint256 totalVotingPower;
+        // Get the voting power of each staking contract
+        for (uint256 i = 0; i < numberStakingContracts; i++) {
+            address staingContractAddress = EMISSIONS_CONTROLLER.stakingContracts(stakingDialIds[i]);
```
Comment (alsco77) [inline contracts/buy-and-make/RevenueBuyBack.sol:175]: Looking up stakingContracts with dialId is incorrect. Position in dials array != position in stakingContracts array. You can just look up the dialData.address to get the address of the staking contract.

File: contracts/buy-and-make/RevenueBuyBack.sol [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+import { Initializable } from "@openzeppelin/contracts/proxy/utils/Initializable.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { IUniswapV3SwapRouter } from "../peripheral/Uniswap/IUniswapV3SwapRouter.sol";
+
+struct RevenueBuyBackConfig {
+    // Percentage of mAssets that can be redeemed for bAssets in basis points (BASSET_PECENTAGE_SCALE).
+    // So if a 2% slippage is allowed, the minRedeemSlippage will be 9800.
+    // A 0.01% slippage is 9999
+    uint16 minBassetPercentage;
```
Comment (alsco77) [inline contracts/buy-and-make/RevenueBuyBack.sol:17]: May want to scale both this var and `minBasset2RewardsPrice` up to some higher amount of 0's in order to accommodate different decimals

File: contracts/emissions/BasicRewardsForwarder.sol [hunk 3]
```diff
@@ -29,13 +31,11 @@ contract BasicRewardsForwarder is
...
 
     /**
-     * @dev Constructor
      * @param _nexus        mStable system Nexus address
      * @param _rewardsToken Token that is being distributed as a reward. eg MTA
      */
     constructor(address _nexus, address _rewardsToken)
-        InitializableRewardsDistributionRecipient(_nexus)
-    {
+        InitializableRewardsDistributionRecipient(_nexus) {
```
Comment (alsco77) [inline contracts/emissions/BasicRewardsForwarder.sol:38]: Husky commitlint still not running or we have mismatching configs 🤔  weird

File: contracts/interfaces/IEmissionsController.sol [hunk 4]
```diff
@@ -0,0 +1,14 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity 0.8.6;
+
+import { IVotes } from "../interfaces/IVotes.sol";
```
Comment (alsco77) [inline contracts/interfaces/IEmissionsController.sol:4]: Unused import

**Review by alsco77 — CHANGES_REQUESTED** [review]

File: contracts/buy-and-make/RevenueBuyBack.sol (hunk 1 above)
Comment (naddison36) [reply contracts/buy-and-make/RevenueBuyBack.sol:175]: Good spot. Thanks

File: contracts/buy-and-make/RevenueBuyBack.sol (hunk 2 above)
Comment (naddison36) [reply contracts/buy-and-make/RevenueBuyBack.sol]: I've change both to use 1e18

File: contracts/emissions/BasicRewardsForwarder.sol (hunk 3 above)
Comment (naddison36) [reply contracts/emissions/BasicRewardsForwarder.sol:38]: I forgot to install husky on my new global node version.

**Pull request merged**